streams/src/main/java/org/apache/kafka/streams/state/RocksDBWindowStore.java

```diff
@@ -0,0 +1,281 @@
+        }
+
+        @Override
+        public void remove() {
+            if (index < iterators.length)
+                iterators[index].remove();
+        }
+
+        public void close() {
+            for (KeyValueIterator<byte[], byte[]> iterator : iterators) {
+                iterator.close();
+            }
+        }
+    }
+
+    private final String name;
+    private final long windowBefore;
+    private final long windowAfter;
+    private final long segmentInterval;
+    private final Segment[] segments;
+    private final Serdes<K, V> serdes;
+    private final SimpleDateFormat formatter;
+
+    private ProcessorContext context;
+    private long currentSegmentId = -1L;
+    private int seqnum = 0;
+
+    public RocksDBWindowStore(String name, long windowBefore, long windowAfter, long retentionPeriod, int numSegments, Serdes<K, V> serdes) {
+        this.name = name;
+        this.windowBefore = windowBefore;
+        this.windowAfter = windowAfter;
+
+        // The retention period must be at least four times as long as the windowSize
+        if ((this.windowBefore + this.windowAfter + 1) * 2 > retentionPeriod)
+            retentionPeriod = (this.windowBefore + this.windowAfter + 1) * 2;
+
+        // The segment interval must be greater than MIN_SEGMENT_INTERVAL
+        this.segmentInterval = Math.max(retentionPeriod / (numSegments - 1), MIN_SEGMENT_INTERVAL);
+
+        this.segments = new Segment[numSegments];
+        this.serdes = serdes;
+
+        // Create a date formatter. Formatted timestamps are used as segment name suffixes
+        this.formatter = new SimpleDateFormat("yyyyMMddHHmm");
+        this.formatter.setTimeZone(new SimpleTimeZone(0, "GMT"));
+    }
+
+    @Override
+    public String name() {
+        return name;
+    }
+
+    @Override
+    public void init(ProcessorContext context) {
+        this.context = context;
+    }
+
+    @Override
+    public boolean persistent() {
+        return true;
+    }
+
+    @Override
+    public void flush() {
+        for (Segment segment : segments) {
+            if (segment != null)
+                segment.flush();
+        }
+    }
+
+    @Override
+    public void close() {
+        for (Segment segment : segments) {
+            if (segment != null)
+                segment.close();
+        }
+    }
+
+    @Override
+    public void put(K key, V value) {
+        putAndReturnInternalKey(key, value);
+    }
+
+    @Override
+    public byte[] putAndReturnInternalKey(K key, V value) {
+        long timestamp = context.timestamp();
+        long segmentId = segmentId(timestamp);
+
+        if (segmentId > currentSegmentId) {
+            // A new segment will be created. Clean up old segments first.
+            currentSegmentId = segmentId;
+            cleanup();
+        }
+
+        // If the record is within the retention period, put it in the store.
+        if (segmentId > currentSegmentId - segments.length) {
```

Comment:
```currentSegmentId - segments.length``` is one less than the smallest valid segment id.

```segmentId``` can be less than ```currentSegmentId```. We do not assume that records come in the chronological order. The range check is still necessary.